Add query genotypes

tpoterba opened the pull request.

cseed commented on src/main/scala/is/hail/variant/VariantDataset.scala:
```diff
@@ -1178,6 +1178,42 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
+    })
+
+    val result = vds.expandWithAll()
+        .map { case (v, va, s, sa, g) => (g, v, va, s, sa) }
```
Can you make the order of the types in makeFunctions match expandWithAll so you don't need to do this?  It allocates a 5-tuple for every genotype!  SAD.

tpoterba replied:
the first element of the tuple is the aggregator type

tpoterba replied:
I don't completely understand this system we've built

cseed commented on src/main/scala/is/hail/expr/FunctionRegistry.scala:
```diff
@@ -459,6 +489,12 @@ object FunctionRegistry {
     bind(name, MethodType(hrt.typ, hru.typ, hrv.typ, hrw.typ), Arity3Aggregator[T, U, V, W, X](hrx.typ, ctor))
   }
 
+  def registerAggregator[T, U, V, W, X, Y, Z](name: String, ctor: (U, V, W, X, Y) => TypedAggregator[Z])
```
I'm confused why you had to add this.  It doesn't look like you register an arity 5 aggregator anywhere.

tpoterba commented:
Addressed comments (Removed the aggregator registry stuff) and made it more better, more faster.  
```text

In [4]: %timeit vds.count(genotypes=True)
1 loop, best of 3: 20.1 s per loop

In [5]: %timeit vds.query_genotypes('gs.fraction(g => g.isCalled)')[0]
1 loop, best of 3: 30.2 s per loop
```

Previous iteration took 40s.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+
+        **Examples**
+
+        >>> gq_hist, t = vds.query_genotypes_typed(
```
Misleading, gq_hist, t are both lists.  What about `[gq_hist], [t] = vds.query...`.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+        >>> gq_hist, t = vds.query_genotypes_typed(
+        ...     'gs.map(g => g.gq).hist(0, 100, 100)')
+
+        See :py:meth:`.query_genotypes` for more information.
```
Shouldn't query_genotypes see this for more information?

tpoterba replied:
This is how we've done the others.  I'll push this in in the same fashion and we can change the others at another time.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+        This method evaluates Hail expressions over genotypes, along with
+        all variant and sample metadata for that genotype. The ``exprs``
+        argument requires either a list of strings or a single string
+        (which will be interpreted as a list with one element).  The method
```
While technically true, the parenthetical seems misleading.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+        all variant and sample metadata for that genotype. The ``exprs``
+        argument requires either a list of strings or a single string
+        (which will be interpreted as a list with one element).  The method
+        returns a list of results (which contains one element if the input
```
and a list of types

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+        - ``global``: global annotations
+        - ``gs`` (*Aggregable[Genotype]*): aggregable of :ref:`genotype`
+
+        Map and filter expressions on this aggregable have the additional
```
Additional seems wrong.  The aggregable namespace does not include the outer namespace.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+        >>> exprs = ['gs.count()', 'gs.filter(g => v.altAllele.isSNP() && g.isHet).count()']
+        >>> results = vds.query_genotypes(exprs)
+
+        :param exprs: one or more query expressions
```
Capital with a period.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+
+        Compute global GQ histogram
+
+        >>> gq_hist = vds.query_genotypes('gs.map(g => g.gq).hist(0, 100, 100)')[0]
```
I increasingly think `[gq_hist] = ...` is more readable.

tpoterba replied:
didn't even realize python had this "unapply" syntax.  This is great.

cseed commented on python/hail/dataset.py:
```diff
@@ -2940,6 +2940,84 @@ def query_variants(self, exprs):
+
+        >>> call_rate = vds.query_genotypes('gs.fraction(g => g.isCalled)')[0]
+
+        :param exprs: one or more query expressions
```
Same.

cseed commented on src/main/scala/is/hail/expr/Fun.scala:
```diff
@@ -58,6 +58,24 @@ case class Arity3Aggregator[T, U, V, W, X](retType: Type, ctor: (U, V, W) => Typ
   }
 }
 
+case class Arity5Aggregator[T, U, V, W, X, Y, Z](retType: Type, ctor: (U, V, W, X, Y) => TypedAggregator[Z]) extends Fun {
```
Is this still needed?  It doesn't seem used.

cseed replied:
You missed this.

cseed commented on src/main/scala/is/hail/variant/VariantDataset.scala:
```diff
@@ -1186,6 +1187,58 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
+        }
+      }
+      Iterator(zv)
+    }.fold(zVal.map(_.copy()))(combOp)
```
Nice and elegant.  Is this not what you wanted to write?  On Slack you mentioned a problem with copy vs serialize/deserialize, but I don't see the problem.

tpoterba replied:
This is the result after fixing that!  I do want to note, though, that the following code does _NOT_ work:

```scala
vds.rdd.treeAggregate(zVal)(seqOp, combOp)
```

...because the treeAggregate does a serialize/deserialize on the zero value, which contain array references.  This is a bit confusing, but I confirmed it a few times.

tpoterba commented:
Back to you.

cseed commented on src/main/scala/is/hail/expr/FunctionRegistry.scala:
```diff
@@ -1482,14 +1482,14 @@ object FunctionRegistry {
 
     >>> from collections import Counter
-    >>> counter = Counter(vds.query_variants('variants.flatMap(v => v.altAlleles).counter()')[0])
+    >>> counter = Counter(vds.query_variants(['variants.flatMap(v => v.altAlleles).counter()'])[0])
```
Put brackets on the LHS.

cseed commented on src/main/scala/is/hail/expr/FunctionRegistry.scala:
```diff
@@ -1565,7 +1565,7 @@ object FunctionRegistry {
     Compute global GQ-distribution:
 
     >>> gq_dist = (vds.annotate_variants_expr('va.gqHist = gs.map(g => g.gq).hist(0, 100, 20)')
```
Brackets on the LHS.
